src/apps/zeromq/gpsd_client/gpsd_client.cpp

```diff
@@ -0,0 +1,425 @@
+// You should have received a copy of the GNU General Public License
+// along with Goby.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <gps.h>
+#include <libgpsmm.h>
+
+#include <boost/units/systems/angle/degrees.hpp>
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+
+#include "goby/util/thirdparty/nlohmann/json.hpp"
+#include "gpsd_client.h"
+
+#include "goby/time.h"
+
+#include "goby/middleware/gpsd/groups.h"
+#include "goby/zeromq/application/single_thread.h"
+
+using goby::glog;
+
+namespace
+{
+// Minor tweak to handle the time format from GPSD;
+goby::time::SITime parse_time(std::string s)
+{
+    // For reference.
+    // https://gpsd.io/gpsd_json.html
+    // From Documentation:
+    // Time/date stamp in ISO8601 format, UTC. May have a fractional part of up to .001sec precision. May be absent if the mode is not 2D or 3D.
+
+    //std::string date_as_string = data["time"].get<std::string>();
+    // Remove the "Z" from the end of the string because it messes up the
+    // parser.
+
+    std::string work_string(s);
+    work_string.pop_back();
+    boost::posix_time::ptime t(boost::posix_time::from_iso_extended_string(work_string));
+
+    return goby::time::convert<goby::time::SITime>(t);
+}
+} // end of Anonymous namespace
+
+goby::apps::zeromq::GPSDClient::GPSDClient()
+    : goby::zeromq::SingleThreadApplication<protobuf::GPSDConfig>(), publish_all_(false)
+{
+    if (cfg().device_name_size())
+    {
+        for (auto& d_name : cfg().device_name()) { device_list_.insert(d_name); }
+    }
+    else
+    {
+        publish_all_ = true;
+
+        // TODO (Shawn): IS this really what we want to do?
+        glog.is_warn() && glog << "No device configured. We will publish all GPS data."
+                               << std::endl;
+    }
+
+    gpsmm gps_rec(cfg().hostname().c_str(), cfg().port().c_str());
+
+    if (!gps_rec.is_open())
+    {
+        // Check constructor success.
+        glog.is_die() && glog << "Could not connect to GPSD at " << cfg().hostname()
+                              << " port: " << cfg().port() << std::endl;
+    }
+
+    // compare devices from our config to what gps_rec has in it's list.
+    if (gps_rec.stream(WATCH_ENABLE | WATCH_JSON) == NULL)
+    {
+        glog.is_die() && glog << "No GPSD running.\n";
+    }
+
+    while (true)
+    {
+        struct gps_data_t* gps_data;
+
+        if (!gps_rec.waiting(1000))
+        {
+            // Notify data missing?
+            continue;
+        }
+
+        if ((gps_data = gps_rec.read()) == NULL)
+        {
+            glog.is_die() && glog << "Read error..\n";
+        }
+        else
+        {
+            try
+            {
+                using json = nlohmann::json;
+
+                json json_data = json::parse(gps_rec.data());
+
+                if (json_data.contains("class"))
+                {
+                    auto& j_class = json_data["class"];
+                    if (j_class == "TPV")
+                    {
```

Comment:
Remove curly braces for one line `if` statements? Quite a few of these and it really grows the code.

Reply:
It's a habit that rolled over from previous code style.  I can take them out.

Reply:
Yeah since I like the Allman brace style I tend to exclude braces on one line clauses to avoid lots of extra lines of code.